packages/teleport-generator-core/src/validator/utils.ts

```diff
@@ -0,0 +1,186 @@
+              node.content.meta.iteratorName
+            }"`
+            errors.push(errorMsg)
+          }
+          if (childNode.content.id && !node.content.meta.useIndex) {
+            const errorMsg = `\nIndex variable is used but the "useIndex" meta information is false.`
+            errors.push(errorMsg)
+          }
+        }
+      })
+    }
+  })
+  return { errors }
+}
+
+// All referenced props and states should be previously defined in the
+// "propDefinitions" and "stateDefinitions" sections
+// If props or states are defined and not used, a warning witll be displayed
+export const checkDynamicDefinitions = (input: any) => {
+  const definedKeys = Object.keys(input.propDefinitions || {}).concat(
+    Object.keys(input.stateDefinitions || {})
+  )
+  const usedKeys = []
+  const errors = []
+
+  traverseNodes(input.node, (node) => {
+    if (
+      node.type === 'dynamic' &&
+      (node.content.referenceType === 'prop' || node.content.referenceType === 'state')
+    ) {
+      if (!definedKeys.includes(node.content.id)) {
+        const errorMsg = `"${
+          node.content.id
+        }" is used but not defined. Please add it in definitions`
+        errors.push(errorMsg)
+      }
+      usedKeys.push(node.content.id)
+    }
+  })
+
+  definedKeys
+    .filter((x) => !usedKeys.includes(x))
+    .map((diff) => console.warn(`"${diff}" is defined but it is not used in the UIDL.`))
+
+  return { errors }
+}
+
+// A projectUIDL must contain "route" key
+export const checkRouteDefinition = (input: ProjectUIDL) => {
+  const errors = []
+
+  const keys = Object.keys(input.root.stateDefinitions || {})
+  if (!keys.includes('route')) {
+    const errorMsg = 'Route is not defined in stateDefinitions'
+    errors.push(errorMsg)
+  }
+  return { errors }
+}
+
+// All referenced components inside of the projectUIDL should be defined
+// in the component section
+export const checkComponentExistence = (input: ProjectUIDL) => {
+  const errors = []
+  const dependencies = Object.keys(input.components)
+
+  traverseElements(input.root.node, (element) => {
+    if (
+      element.dependency &&
+      element.dependency.type === 'local' &&
+      !dependencies.includes(element.elementType)
+    ) {
+      const errorMsg = `\nThe component "${
+        element.elementType
+      }" is not defined in the UIDL's component section.`
+      errors.push(errorMsg)
+    }
+  })
+  return { errors }
+}
+
+// All components should have the same key as the value of their name key
+// Example:
+//  "components": {
+//     "OneComponent": {
+//       "name": "OneComponent",
+//    ..
+//      }
+//    }
+export const checkComponentNaming = (input: ProjectUIDL) => {
+  const errors = []
+  const namesUsed = Object.keys(input.components)
+
+  namesUsed
+    .filter((name) => input.components[name].name !== name)
+    .map((diff) => {
+      const errorMsg = `\nThe following dependencies have different name than their key: ${diff}`
```

Comment:
shouldn't the message be displayed only once and the the list of components? comma separated?

Reply:
also, I would name them "components", not "dependencies", not to be confused with the dependencies inside elementNodes